packages/core/src/middleware/koa-log.ts

```diff
@@ -41,14 +32,30 @@ export default function koaLog<StateT, ContextT, ResponseBodyT>(): MiddlewareTyp
   ResponseBodyT
 > {
   return async (ctx, next) => {
-    ctx.log = {};
+    // eslint-disable-next-line @silverhand/fp/no-let
+    let logType: Optional<LogType>;
+    // eslint-disable-next-line @silverhand/fp/no-let
+    let logPayload: LogPayload = {};
+
+    ctx.log = <T extends LogType, P extends LogPayloads[T]>(type: T, payload: P) => {
+      if (!logType) {
+        // eslint-disable-next-line @silverhand/fp/no-mutation
+        logType = type; // Use first initialized log type
+      }
+      // eslint-disable-next-line @silverhand/fp/no-mutation
+      logPayload = deepmerge(logPayload, payload);
+    };
```

Comment:
you should check if the `type` parameter is equal to `logType`, otherwise the type restriction here is kind of useless if we assign 2 different log types with different payload sequentially